src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -197,49 +188,65 @@ class FastSync(
         handleRequestFailure(assignedHandlers(ref), ref, "Unexpected error")
     }
 
-    def waitingForPivotBlockUpdate(processState: FinalBlockProcessingResult): Receive = handleCommonMessages orElse {
+    def askForPivotBlockUpdate(updateReason: PivotBlockUpdateReason): Unit = {
+      syncState = syncState.copy(updatingPivotBlock = true)
+      log.info("Asking for new pivot block")
+      val pivotBlockSelector = {
+        context.actorOf(
+          PivotBlockSelector.props(etcPeerManager, peerEventBus, syncConfig, scheduler, context.self)
+        )
+      }
+      pivotBlockSelector ! PivotBlockSelector.SelectPivotBlock
+      context become waitingForPivotBlockUpdate(updateReason)
+    }
+
+    def reScheduleAskForNewPivot(updateReason: PivotBlockUpdateReason): Unit = {
+      syncState = syncState.copy(pivotBlockUpdateFailures = syncState.pivotBlockUpdateFailures + 1)
+      scheduler.scheduleOnce(syncConfig.pivotBlockReScheduleInterval, self, UpdatePivotBlock(updateReason))
+    }
+
+    def waitingForPivotBlockUpdate(updateReason: PivotBlockUpdateReason): Receive = handleCommonMessages orElse {
       case PivotBlockSelector.Result(pivotBlockHeader) =>
         log.info(s"New pivot block with number ${pivotBlockHeader.number} received")
         if (pivotBlockHeader.number >= syncState.pivotBlock.number) {
-          updatePivotSyncState(processState, pivotBlockHeader)
-          syncState = syncState.copy(updatingPivotBlock = false)
-          context become this.receive
-          processSyncing()
+          if (pivotBlockHeader.number == syncState.pivotBlock.number && updateReason.nodeRestart) {
+            // it can happen after quick node restart than pivot block has not changed in the network. To keep whole
+            // fast sync machinery running as expected we need to make sure that we will receive better pivot than current
+            log.info("Received stale pivot after restart, asking for new pivot")
+            reScheduleAskForNewPivot(updateReason)
+          } else {
+            updatePivotSyncState(updateReason, pivotBlockHeader)
+            syncState = syncState.copy(updatingPivotBlock = false)
```

Comment:
I think `syncState = syncState.copy(updatingPivotBlock = false)` should be done in `updatePivotSyncState` method